src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -274,15 +642,310 @@ TryDropShard(GroupShardPlacement *placement)
 									   dropQuery->data);
 
 	/* remove the shard from the node */
-	bool success =
-		SendOptionalCommandListToWorkerOutsideTransaction(shardPlacement->nodeName,
-														  shardPlacement->nodePort,
-														  NULL, dropCommandList);
-	if (success)
+	bool success = SendOptionalCommandListToWorkerOutsideTransaction(nodeName,
+																	 nodePort,
+																	 NULL,
+																	 dropCommandList);
+
+	return success;
+}
+
+
+/*
+ * GetNextOperationId allocates and returns a unique operationId for an operation
+ * requiring potential cleanup. This allocation occurs both in shared memory and
+ * in write ahead logs; writing to logs avoids the risk of having operationId collisions.
+ */
+static OperationId
+GetNextOperationId()
+{
+	OperationId operationdId = INVALID_OPERATION_ID;
+
+	/*
+	 * In regression tests, we would like to generate operation IDs consistently
+	 * even if the tests run in parallel. Instead of the sequence, we can use
+	 * the next_operation_id GUC to specify which operation ID the current session should
+	 * generate next. The GUC is automatically increased by 1 every time a new
+	 * operation ID is generated.
+	 */
+	if (NextOperationId > 0)
 	{
-		/* delete the actual placement */
-		DeleteShardPlacementRow(placement->placementId);
+		operationdId = NextOperationId;
+		NextOperationId += 1;
+
+		return operationdId;
 	}
 
-	return success;
+	/* Generate sequence using a subtransaction. else we can hold replication slot creation for operations */
+	StringInfo sequenceName = makeStringInfo();
+	appendStringInfo(sequenceName, "%s.%s",
+					 PG_CATALOG,
+					 OPERATIONID_SEQUENCE_NAME);
+
+	StringInfo nextValueCommand = makeStringInfo();
+	appendStringInfo(nextValueCommand, "SELECT nextval(%s);",
+					 quote_literal_cstr(sequenceName->data));
+
+	int connectionFlag = FORCE_NEW_CONNECTION;
+	MultiConnection *connection = GetNodeUserDatabaseConnection(connectionFlag,
+																LocalHostName,
+																PostPortNumber,
+																CitusExtensionOwnerName(),
+																get_database_name(
+																	MyDatabaseId));
+
+	PGresult *result = NULL;
+	int queryResult = ExecuteOptionalRemoteCommand(connection, nextValueCommand->data,
```

Comment:
looking through the code a bit, it seems our usual pattern for sending a command and getting the result is:
```
    int querySent = SendRemoteCommand(connection, query);
    if (querySent == 0)
    {
        ReportConnectionError(connection, ERROR);
    }

    PGresult *result = GetRemoteCommandResult(connection, raiseInterrupts);
    if (!IsResponseOK(result))
    {
        ReportResultError(superuserConnection, result, ERROR);
    }

    ... do stuff with result...

    PQclear(result);
    ForgetResults(superuserConnection);
```

it seems preferable to follow that pattern, e.g. the current implementation leaks memory since it does not call PQclear

Reply:
The 'ExecuteOptionalRemoteCommand' does the same and has SendRemoteCommand+GetRemoteCommandResult as well right? I will make sure we call PQclear(result)+ ForgetResults(superuserConnection);

Reply:
Updated.